### ISSUE TITLE
Allow setting multiple render targets using Painter (Starling 2.0)

### ISSUE THREAD

**Issue by Varnius** [Gamua/Starling-Framework#792]
I'd like to suggest modifying Painter and RenderState to support setting multiple render targets.

In my case, I do need to specify all 4 RTs and doing this directly on context object would probably beat the purpose of Painter state mechanism. Also, I'm not sure how I would handle applyClipRect myself, since it should be done automatically by Painter.

So why not make this
```actionscript
public function setRenderTarget(target:Texture, antiAlias:int=0):void
```
to look like this:
```actionscript
public function setRenderTarget(target:Texture, antiAlias:int=0, colorOutputIndex:int=0):void
```

**Comment by PrimaryFeather** [comment]
You're perfectly right, this needs to be done at Painter level. In fact, I planned to do this right away, but then postponed the change to get some other things done first. I'll make that change soon!

**Comment by Varnius** [comment]
Good to hear, thanks!

**Comment by Varnius** [comment]
By the way, what about enableDepthAndStencil param?
I think it should be included as an optional param for finer grained control over how the rendering pipeline works because in most cases only single RT will need depth buffer enabled.

**Comment by PrimaryFeather** [comment]
Mhm, that would also make sense. I have to be very careful with how I implement all of this, though, because the state stack is an extremely sensitive area of the code, performance wise. I'll get back on you! =)

**Comment by PrimaryFeather** [comment]
I just had a look at this, and I'd like to discuss something with you — I have never written any shaders utilizing multiple render targets before, so I'm glad about any input from you! :smile:

I realize that all texture targets need to have the same format and size, and you need to manually assign the output color in the fragment shader (i.e. copy color to `fo0`, `fo1`, etc).

Thus, I'm no longer sure if this really needs to / should be done on the render state. Because the state is supposed to affect all subsequent draw calls: changing its alpha will change the alpha of all subsequent vertices; changing its render target will make *any* fragment shaders draw their output into the given texture.

However, if I assign, say, render targets 0-2 on the context (painter), I guess that all fragment shaders that don't copy any values to `fo1` and `fo2` will simply fail, right? (Or, best case: won't do anything with that data.)

In other words, whenever you use multiple render targets, you also need to use specific fragment shaders that accept that data. That's always coupled.

Now, what to do about the clipRect? Well, since all texture target need to use an identical layout, anyway, it suffices to assign the *first* one (`fo0`) to `painter.setRenderTarget()`. This will take care of the clip rect. Then, inside your "Effect" subclass, you'd assign the other targets directly on the context in `beforeDraw` and clear them afterwards in `afterDraw`.

It's not that I'd be too lazy to implement this, I'm just honestly not sure anymore that it's the best idea to add that on the painter. If setting multiple render targets is always coupled with using special fragment programs, the effect class just seems to be the better fit, because otherwise the painter would needlessly have to work with this data *every time* somebody pushes or pops a state (which happens for *any* display object). Just as the effect assigns the right number of textures for the fragment program.

What do you say — does that make sense? Or am I overlooking something?

**Comment by Varnius** [comment]
You have a point. Actually I thought about setting only first RT for the state and handling all others directly on the context but figured that would not be very clean.
And yes, shaders that do not output to all RTs when that is required will fail.
 I'll do some more digging and get back to you. :ghost:

**Comment by PrimaryFeather** [comment]
Awesome, thanks for the feedback! As for the depth and stencil parameter, that should definitely be a part of the state; I'll add that next week.

**Comment by PrimaryFeather** [comment]
I just added that `enableDepthAndStencil` parameter. I left the issue open for now, though, since it was about providing multiple targets, originally. Feel free to close the issue once you've finished your experiments and can approve that we don't need that change!

**Comment by b005t3r** [comment]
In my opinion, you should be able to set multiple render targets using Painter instead of being forced to use underlying Context. I can't remember exactly what was it, but we already had a similar discussion about RenderSupport and in the end we agreed that RenderSupport is supposed to be a facade and all you do, should be done using RenderSupport's methods instead of using Context3D directly.

> However, if I assign, say, render targets 0-2 on the context (painter), I guess that all fragment shaders that don't copy any values to fo1 and fo2 will simply fail, right?

No sure if I'm missing something here, but if you use multiple render targets you are obliged to clean up after doing your rendering and unset them, right?

> It's probably more similar to texture samplers: if the fragment shader uses texture samplers, the associated effect will assign them.

Should be a part of Painter as well and I think a good point here is performance optimization. Currently, even if you manage to fit your entire app resources into a single texture, you're still going to have code which binds that texture, renders and unbinds it (because there's no way of knowing the texture you want to sample from is already set, using Context3D), but you should be able to bind the texture once and never have to unbind it. This can easily be done when forwarding calls such as ```setTextureAt()``` via Painter.

For the software architecture point of view it is also a better approach not to split similar functionalities between two classes, especially if one is a facade for the other - how am I supposed to know which Context3D methods I can call safely (being sure I'm not breaking what Painter does) and which I can't? The only safe way when writing a client code is to rely on Painter method's only, because their internals might change in the future and this might break my code which uses Context3D directly.

TL;DR: If Painter is a facade for Context3D methods, forbid using Context3D directly (it should be used internally by Painter only) and add necessary methods in Painter.

**Comment by PrimaryFeather** [comment]
Yes, we had such a discussion a while ago, and that was in part responsible for my coming up with this new "Painter" and "RenderState" design. If I remember correctly, your rendered something into a texture, and this modified the render target, which you could not restore in Starling 1.x, because you didn't have access to the `RenderUtil` outside the render methods. Now, you can access the `Painter` from anywhere via `Starling.painter`, which (together with the state stack) solves this problem.

However, about your suggestion: if you look at the new classes (`Painter` and `RenderState`) more closely, you'll see that they actually only wrap / replace very few methods of `Context3D`. Most of the methods are strictly Starling-related states that have no equivalent in the context; or, if they have an equivalent, most can be safely called on the context, just as well.

The exceptions are: `painter.present()`, `state.culling`, `state.clipRect`, `state.blendMode` and `state.renderTarget`. 

Indeed: for those methods, it's always wise to call them on the `Painter` or `State` instead of the context. In the cases `clipRect` and `blendMode`, you have no other choice, anyway, because they have no 1:1 equivalent on the context, but are abstractions that just happen to use the context internally.

Now, I agree: if I added all other context methods to the state, it would be extremely convenient for developers, and I'd love to be able to do that. In fact, that was the original plan. ;-) 

However, the state would also become *very* large; thus, pushing and popping it would become very expensive — so much so that it would no longer be feasible to use it at all. Remember, *every* display object has to make a "push" and a "pop" once per frame! So I had to find a more pragmatic solution, and wrapped only the handful critical properties mentioned above.

> [Texture samplers] should be a part of Painter as well and I think a good point here is performance optimization.

Actually, if you think about this, it wouldn't help much in reality: the whole architecture with the batches grouping similar geometry together will make sure that the texture changes in almost each render call, anyway. Not much to gain here.

> Not sure if I'm missing something here, but if you use multiple render targets you are obliged to clean up after doing your rendering and unset them, right?

Exactly! And if one has to clean up that stuff anyway (i.e. it's not being shared with other display objects, like `clipRect`, `alpha`, etc. would be), it doesn't need to be part of the state stack.

All in all, I think it's going to seem quite intuitive when you write your first own mesh style:

When writing the `MeshEffect` subclass, you'll make your modifications to the context in the `beforeDraw` method and undo your changes in the `afterDraw` method. You don't access either the `Painter`, nor the `state` in those methods! You could, technically, because you can always access them via `Starling.painter`, but they are not passed to these methods. Everything you do here is just for the sake of this one object that is being rendered; it doesn't make sense to "leak" out any context settings; you clean up after yourself!

However, in the "MeshStyle" class, you'll typically access the state, that's why the `updateEffect` method has a `RenderState` parameter. Accessing transformation matrices, changing clipping, etc., can all be done here, and will automatically be undone later by the next "popState" call. There's no `after...` method here, so it should be intuitive that there's no clean-up to be done.

Finally, for those state settings that affect a whole subtree of display objects (e.g. `clipRect`, `blendMode`, `alpha`, etc.), this is all done by Starling, and you don't need to access those properties typically.

Makes sense?

**Comment by b005t3r** [comment]
it does make sense, thanks for the extensive explanation, Daniel!

I understand you had to take the "bets performance" route instead of the "kosher OOP" one :smiley: I still didn't have time to dive into 2.0 internals and try things out, but what wrote sums up to a well thought-through solution and I can probably live with that :smiley:

**Comment by PrimaryFeather** [comment]
Perfect, I'm glad you're with me! Let me know when you're starting to play around with it, too, and anything comes up. :smile:

**Comment by b005t3r** [comment]
Will do for sure. And when do you plan to make the 2.0 official?

**Comment by Varnius** [comment]
Yup, your thoughts really make sense, Daniel. I still haven't had time to spend on rewriting my extension but I'm hoping to find some, I'll post more feedback then. :)

**Comment by PrimaryFeather** [comment]
Thanks, guys! :smile:

What's still missing before I can release the version is the updated filter API, as well as "Scale3" and "Scale9" textures. My hope is that I can release in February.
